python/tank_vendor/shotgun_deploy/configuration.py

```diff
@@ -0,0 +1,983 @@
+    config_path_win = append_folder_to_path("win32", win_path, "config")
+    config_path_linux = append_folder_to_path("linux2", linux_path, "config")
+    config_path_mac = append_folder_to_path("darwin", mac_path, "config")
+
+    config_location = {
+        "type": "path",
+        "linux_path": config_path_linux,
+        "mac_path": config_path_mac,
+        "windows_path": config_path_win
+    }
+
+    config_descriptor = create_descriptor(
+        sg,
+        Descriptor.CONFIG,
+        config_location,
+        fallback_roots=bundle_cache_fallback_paths
+    )
+
+    return ManagedConfiguration(
+        sg,
+        config_descriptor,
+        project_id,
+        pipeline_config_id,
+        namespace,
+        bundle_cache_fallback_paths,
+        use_global_bundle_cache,
+        config_root)
+
+
+class Configuration(object):
+    """
+    An abstraction around a toolkit configuration.
+
+    The configuration is identified by a ConfigurationDescriptor
+    object and may or may not exist on disk.
+    """
+
+    (LOCAL_CFG_UP_TO_DATE, LOCAL_CFG_MISSING, LOCAL_CFG_OLD, LOCAL_CFG_INVALID) = range(4)
+
+    def __init__(
+            self,
+            sg,
+            descriptor,
+            project_id,
+            pipeline_config_id,
+            namespace,
+            bundle_cache_fallback_paths
+    ):
+        """
+        Constructor.
+
+        :param sg: Shotgun API instance
+        :param descriptor: ConfigDescriptor for the associated config
+        :param project_id: Project id for the shotgun project associated with the
+                           configuration. For a site-level configuration, this
+                           can be set to None.
+        :param pipeline_config_id: Pipeline Configuration id for the shotgun
+                                   pipeline config id associated. If a config does
+                                   not have an associated entity in Shotgun, this
+                                   should be set to None.
+        :param namespace: name space string, typically one short word,
+                          e.g. 'maya', 'rv', 'desktop'.
+        :param bundle_cache_fallback_paths: List of additional paths where apps are cached.
+        """
+        self._sg_connection = sg
+        self._descriptor = descriptor
+        self._project_id = project_id
+        self._pipeline_config_id = pipeline_config_id
+        self._namespace = namespace
+        self._bundle_cache_fallback_paths = bundle_cache_fallback_paths
+
+    def __repr__(self):
+        return "<Config with id %s, project id %s and base %s>" % (
+            self._pipeline_config_id,
+            self._project_id,
+            self._descriptor
+        )
+
+    def get_descriptor(self):
+        """
+        Returns the descriptor that is associated with this configuration
+
+        :return: ConfigDescriptor Object
+        """
+        return self._descriptor
+
+    def get_path(self, platform=sys.platform):
```

Comment:
I wish we could leave Python 2.5 behind and use the [abc](https://docs.python.org/2/library/abc.html) module.